solidity/contracts/bridge/RedemptionWatchtower.sol

```diff
@@ -0,0 +1,469 @@
+    ///         from the moment when the redemption request was created.
+    ///         Value in seconds.
+    uint32 public levelTwoDelay;
+
+    event WatchtowerEnabled(uint32 enabledAt, address manager);
+
+    event GuardianAdded(address indexed guardian);
+
+    event GuardianRemoved(address indexed guardian);
+
+    event ObjectionRaised(
+        uint256 indexed redemptionKey,
+        address indexed guardian
+    );
+
+    event VetoPeriodCheckOmitted(uint256 indexed redemptionKey);
+
+    event VetoFinalized(uint256 indexed redemptionKey);
+
+    event WatchtowerParametersUpdated(
+        uint32 watchtowerLifetime,
+        uint64 vetoPenaltyFeeDivisor,
+        uint32 vetoFreezePeriod,
+        uint32 defaultDelay,
+        uint32 levelOneDelay,
+        uint32 levelTwoDelay
+    );
+
+    modifier onlyManager() {
+        require(msg.sender == manager, "Caller is not watchtower manager");
+        _;
+    }
+
+    modifier onlyGuardian() {
+        require(isGuardian[msg.sender], "Caller is not guardian");
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(Bridge _bridge) external initializer {
+        __Ownable_init();
+
+        bridge = _bridge;
+        (bank, , , ) = _bridge.contractReferences();
+
+        watchtowerLifetime = 18 * 30 days; // 18 months
+        vetoPenaltyFeeDivisor = 1; // 100% as initial penalty fee
+        vetoFreezePeriod = 30 days; // 1 month
+        defaultDelay = 2 hours;
+        levelOneDelay = 8 hours;
+        levelTwoDelay = 24 hours;
+    }
+
+    /// @notice Enables the redemption watchtower and veto mechanism.
+    /// @param _manager Address of the watchtower manager.
+    /// @param _guardians List of initial guardian addresses.
+    /// @dev Requirements:
+    ///      - The caller must be the owner,
+    ///      - Watchtower must not be enabled already,
+    ///      - Manager address must not be zero.
+    function enableWatchtower(address _manager, address[] calldata _guardians)
+        external
+        onlyOwner
+    {
+        require(watchtowerEnabledAt == 0, "Already enabled");
+
+        require(_manager != address(0), "Manager address must not be 0x0");
+        manager = _manager;
+
+        for (uint256 i = 0; i < _guardians.length; i++) {
+            _addGuardian(_guardians[i]);
+        }
+
+        /* solhint-disable-next-line not-rely-on-time */
+        uint32 enabledAt = uint32(block.timestamp);
+        watchtowerEnabledAt = enabledAt;
+
+        emit WatchtowerEnabled(enabledAt, _manager);
+    }
+
+    /// @notice Adds a redemption guardian
+    /// @param guardian Address of the guardian to add.
+    /// @dev Requirements:
+    ///      - The caller must be the watchtower manager,
+    ///      - The guardian must not already exist.
+    function addGuardian(address guardian) external onlyManager {
```

Comment:
what happens if the Watchtower is not yet enabled and a Manager has not yet been set?

Reply:
In that case, `addGuardian` cannot be called by anyone. There are no guardians in the system and the veto mechanism is not functioning. The contract is designed in a way that `enableWatchtower` must be called first.

The deployment scenario will be as follows:
1. Deploy the new implementation of the `Bridge` contract and set the transparent proxy to point there
2. Deploy the `RedemptionWatchtower` contract
3. Threshold Council calls `enableWatchtower` with `manager` being the [Governor Bravo `TimelockController`](https://etherscan.io/address/0x87F005317692D05BAA4193AB0c961c69e175f45f) and some initial `guardians`. 
4. Threshold Council attaches the `RedemptionWatchtower` to the `Bridge` via `setRedemptionWatchtower`

Reply:
Can you describe such a scenario in points? I'm not sure if this is possible before calling `enableWatchtower`.

Reply:
If i'm paying careful attention, I will see the `RedemptionWatchtower` being deployed. Soon after (but not instant right?), the council will call `enableWatchtower` with a bunch of initial config, which includes setting the `manager`. My question is, during that small window with no `manager` set, is a malicious actor able to call this `addGuardian` function successfully?

Reply:
Ok I see, thanks for clarification! I think it's not possible because before calling `enableWatchtower` the `manager` will be `address(0)`. When someone calls `addGuardian`, we will jump to the `onlyManager` modifier that will check `msg.sender != address(0)` and revert. I added a test case to prove that: 6b0a76ab

Reply:
ok yeah, I can see how the modifier handles this. Thanks for adding the test though